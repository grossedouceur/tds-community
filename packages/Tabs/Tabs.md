### Usage criteria

- Use `Tabs` according to props data

```jsx
<Tabs
  tabIndex={0}
  tabsEntries={[
    {
      entryTitle: 'Modal - Channels - Tabs - Essentials copy',
      tabTitle: 'Essentials',
      tabContent: [
        {
          entryTitle: 'Essentials Modal - Channels - Tabs - Essentials - Row copy',
          hasGutter: false,
          horizontalAlign: 'start',
        },
      ],
    },
    {
      entryTitle: 'Modal - Channels - Tabs - Essentials copy',
      tabTitle: 'Premium',
      tabContent: [
        {
          entryTitle: 'Premium modal - Channels - Tabs - Essentials - Row copy',
          hasGutter: false,
          horizontalAlign: 'start',
        },
      ],
    },
    {
      entryTitle: 'Modal - channels - Tabs - Essentials copy',
      tabTitle: 'Theme packs',
      tabContent: [
        {
          entryTitle: 'Theme modal - Channels - Tabs - Essentials - Row copy',
          hasGutter: false,
          horizontalAlign: 'start',
        },
      ],
    },
  ]}
/>
```
