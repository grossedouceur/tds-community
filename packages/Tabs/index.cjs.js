const Tabs = require('./dist/index.cjs')

module.exports = Tabs
