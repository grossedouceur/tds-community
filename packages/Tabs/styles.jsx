import styled, { css } from 'styled-components'

export const TabsContainer = styled.div`
  position: relative;

  .react-tabs {
    overflow-x: hidden;
  }

  .react-tabs__tab-list {
    transition: 0.9s all ease;
    transform: translate3d(${props => props.positionToMove}px, 0px, 0px);
    white-space: nowrap;
  }

  .react-tabs__tab {
    display: inline-block;
    cursor: pointer;
    padding: 0 25px;
    min-width: 44px;
    text-align: center;
    position: relative;

    &:first-child {
      padding-left: 0;
    }
    &:last-child {
      padding-right: 0;
    }

    &:hover {
      span {
        text-shadow: 0px 0px 1px #2a2c2e;
        border-bottom: 4px solid #d8d8d8;
      }
    }

    &:active,
    &.react-tabs__tab--selected {
      span {
        text-shadow: 0px 0px 1px #2a2c2e;
        border-bottom: 4px solid #71747a;
      }
    }
  }

  .react-tabs__tab-panel {
    display: none;

    &.react-tabs__tab-panel--selected {
      display: block;
    }
  }
`

export const TabListContainer = styled.div`
  border-bottom: 1px solid #d8d8d8;
  position: relative;

  &:after {
    content: '';
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    height: 32px;
    background-image: radial-gradient(
      at center top,
      rgba(150, 150, 150, 0.1) 0%,
      rgba(0, 0, 0, 0) 70%
    );
    padding: 0px;
    margin: 0px;
    border-width: 0px;
  }
`

export const TabLabel = styled.span`
  height: 100%;
  display: block;
  padding-bottom: 7px;
  border-bottom: 4px solid transparent;
`

export const TabArrows = styled.div`
  position: absolute;
  top: 0;
  background: #fff;
  z-index: 10;
  cursor: pointer;

  ${props =>
    props.direction === 'left' &&
    css`
      left: 0;
      border-right: 1px solid #d8d8d8;
    `};

  ${props =>
    props.direction === 'right' &&
    css`
      right: 0;
      border-left: 1px solid #d8d8d8;
    `};

  &:after {
    content: '';
    position: absolute;
    left: 0;
    bottom: -5px;
    width: 100%;
    height: 5px;
    background: #fff;
  }
`
