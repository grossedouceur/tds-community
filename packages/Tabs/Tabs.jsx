import React, { useRef, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import FlexGrid from '@tds/core-flex-grid'
import { safeRest } from '@tds/util-helpers'
import { ChevronRight, ChevronLeft } from '@tds/core-interactive-icon'
import { Tab, Tabs as ReactTabs, TabList, TabPanel } from 'react-tabs'
import { TabsContainer, TabListContainer, TabLabel, TabArrows } from './styles'
import hash from './hash'
import Panel from './Panel/Panel'
/**
 * @version ./package.json
 * @visibleName Tabs (beta)
 */

const Tabs = props => {
  const tabsRoot = useRef()
  const [tabsContainerWidth, setTabsContainerWidth] = useState()
  const [tabsTranslatePosition, setTabsTranslatePosition] = useState(0)
  const [isScrollEnabled, setScrollEnabled] = useState(false)
  const [isLeftArrow, setIsLeftArrow] = useState(false)
  const [isRightArrow, setIsRightArrow] = useState(false)
  const [current, setCurrent] = useState(0)
  const fullWithRef = useRef(0)
  const { children, ...rest } = props

  const currentIndex = i => {
    setCurrent(i)
  }
  const _nodes = new Map()
  const getTabsWidth = () => {
    let totalTabsWidth = 0
    Object.keys(_nodes).forEach(key => {
      const value = _nodes[key]
      if (value && value.node && value.node.offsetWidth) {
        totalTabsWidth += value.node.offsetWidth
      }
    })

    if (tabsRoot.current.offsetWidth < totalTabsWidth) {
      setTabsContainerWidth(`${totalTabsWidth}px`)
      fullWithRef.current = totalTabsWidth
      setIsRightArrow(true)
      return setScrollEnabled(true)
    }
    return setScrollEnabled(false)
  }
  const setRef = (tabRef, i) => {
    _nodes[i] = tabRef
  }
  const mapTabs = () => {
    if (props.children.length > 0) {
      return props.children.map((tab, i) => {
        return (
          <Tab ref={tabRef => setRef(tabRef, i)} key={hash(i)} onClick={() => currentIndex(i)}>
            <TabLabel>{tab.props.heading}</TabLabel>
          </Tab>
        )
      })
    }
    return ''
  }

  const mapTabContent = () => {
    if (props.children.length > 0) {
      return props.children.map((tab, i) => {
        return (
          <TabPanel key={hash(i)}>
            <FlexGrid>
              <FlexGrid.Row>
                <FlexGrid.Col xs={12}>
                  <Panel {...rest}>{children[current]}</Panel>
                </FlexGrid.Col>
              </FlexGrid.Row>
            </FlexGrid>
          </TabPanel>
        )
      })
    }
    return ''
  }

  const scrollTabs = direction => {
    let currentPosition = tabsTranslatePosition
    if (direction === 'right') {
      currentPosition -= 100
      setIsLeftArrow(true)
    }
    if (direction === 'left') {
      currentPosition += 100
      setIsRightArrow(true)
    }
    if (
      Math.abs(currentPosition) + tabsRoot.current.offsetWidth >= fullWithRef.current &&
      currentPosition < 0
    ) {
      setIsRightArrow(false)
    }
    if (currentPosition === 0) {
      setIsLeftArrow(false)
    }
    setTabsTranslatePosition(currentPosition)
  }
  useEffect(() => {
    if (typeof window !== 'undefined') {
      setTimeout(() => getTabsWidth(), 100)
    }
  }, [])
  return (
    <TabsContainer {...safeRest(rest)} ref={tabsRoot} positionToMove={tabsTranslatePosition}>
      <FlexGrid gutter={false}>
        <FlexGrid.Row>
          <FlexGrid.Col xs={12}>
            {isLeftArrow && (
              <TabArrows direction="left" onClick={() => scrollTabs('left')}>
                <ChevronLeft variant="basic" />
              </TabArrows>
            )}
            <ReactTabs>
              <TabListContainer isScrollEnabled={isScrollEnabled}>
                <TabList style={{ width: tabsContainerWidth }}>{mapTabs()}</TabList>
              </TabListContainer>
              {mapTabContent()}
            </ReactTabs>
            {isRightArrow && (
              <TabArrows direction="right" onClick={() => scrollTabs('right')}>
                <ChevronRight variant="basic" />
              </TabArrows>
            )}
          </FlexGrid.Col>
        </FlexGrid.Row>
      </FlexGrid>
    </TabsContainer>
  )
}

Tabs.propTypes = {
  /**
   * The tab panels. Must be at least one `<Tabs.Panel />`.
   */
  children: PropTypes.node.isRequired,
}
Tabs.Panel = Panel
export default Tabs
