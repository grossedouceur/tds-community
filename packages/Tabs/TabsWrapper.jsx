import React, { useRef, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import FlexGrid from '@tds/core-flex-grid'
import { ChevronRight, ChevronLeft } from '@tds/core-interactive-icon'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'
import { TabsContainer, TabListContainer, TabLabel, TabArrows } from './styles'
import hash from './hash'
// import TabsElement from './TabsElement'
/**
 * @version ./package.json
 * @visibleName Tabs (beta)
 */

const TabsWrapper = ({ tabsEntries, tabIndex }) => {
  const tabsRoot = useRef()
  const [tabsContainerWidth, setTabsContainerWidth] = useState()
  const [tabsTranslatePosition, setTabsTranslatePosition] = useState(0)
  const [isScrollEnabled, setScrollEnabled] = useState(false)
  const [isLeftArrow, setIsLeftArrow] = useState(false)
  const [isRightArrow, setIsRightArrow] = useState(false)
  const fullWithRef = useRef(0)

  const getTabsWidth = () => {
    const tabsArray = document.querySelectorAll('.react-tabs__tab')
    let totalTabsWidth = 0
    tabsArray.forEach(tab => {
      totalTabsWidth += tab.offsetWidth
    })
    if (tabsRoot.current.offsetWidth < totalTabsWidth) {
      setTabsContainerWidth(`${totalTabsWidth}px`)
      fullWithRef.current = totalTabsWidth
      setIsRightArrow(true)
      return setScrollEnabled(true)
    }
    return setScrollEnabled(false)
  }

  const mapTabs = () => {
    if (tabsEntries.length > 0) {
      return tabsEntries.map(tab => {
        return (
          <Tab key={hash(tab)}>
            <TabLabel>{tab.tabTitle}</TabLabel>
          </Tab>
        )
      })
    }
    return ''
  }

  const mapTabContent = () => {
    if (tabsEntries.length > 0) {
      return tabsEntries.map(tab => {
        return (
          <TabPanel key={hash(tab)}>
            <FlexGrid>
              <FlexGrid.Row>
                <FlexGrid.Col xs={12}>{tab && tab.tabContent[0].entryTitle}</FlexGrid.Col>
              </FlexGrid.Row>
            </FlexGrid>
          </TabPanel>
        )
      })
    }
    return ''
  }

  const scrollTabs = direction => {
    let currentPosition = tabsTranslatePosition
    if (direction === 'right') {
      currentPosition -= 100
      setIsLeftArrow(true)
    }
    if (direction === 'left') {
      currentPosition += 100
      setIsRightArrow(true)
    }
    if (
      Math.abs(currentPosition) + tabsRoot.current.offsetWidth >= fullWithRef.current &&
      currentPosition < 0
    ) {
      setIsRightArrow(false)
    }
    if (currentPosition === 0) {
      setIsLeftArrow(false)
    }
    setTabsTranslatePosition(currentPosition)
  }
  useEffect(() => {
    if (typeof window !== 'undefined') {
      setTimeout(() => getTabsWidth(), 10)
    }
  }, [])
  return (
    <TabsContainer ref={tabsRoot} positionToMove={tabsTranslatePosition}>
      <FlexGrid gutter={false}>
        <FlexGrid.Row>
          <FlexGrid.Col xs={12}>
            {isLeftArrow && (
              <TabArrows direction="left" onClick={() => scrollTabs('left')}>
                <ChevronLeft variant="basic" />
              </TabArrows>
            )}
            <Tabs defaultIndex={tabIndex}>
              <TabListContainer isScrollEnabled={isScrollEnabled}>
                <TabList style={{ width: tabsContainerWidth }}>{mapTabs()}</TabList>
              </TabListContainer>
              {mapTabContent()}
            </Tabs>
            {isRightArrow && (
              <TabArrows direction="right" onClick={() => scrollTabs('right')}>
                <ChevronRight variant="basic" />
              </TabArrows>
            )}
          </FlexGrid.Col>
        </FlexGrid.Row>
      </FlexGrid>
    </TabsContainer>
  )
}

TabsWrapper.propTypes = {
  tabsEntries: PropTypes.array,
  tabIndex: PropTypes.number,
}

TabsWrapper.defaultProps = {
  tabsEntries: [],
  tabIndex: PropTypes.any,
}

export default TabsWrapper
