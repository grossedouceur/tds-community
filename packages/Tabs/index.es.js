import Tabs from './dist/index.es'

export default Tabs
