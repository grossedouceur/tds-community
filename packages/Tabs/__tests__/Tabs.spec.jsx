import React from 'react'
import { shallow } from 'enzyme'

import Tabs from '../Tabs'

describe('Tabs', () => {
  const doShallow = (props = {}) => shallow(<Tabs {...props} />)

  it('renders', () => {
    const tabs = doShallow()

    expect(tabs).toMatchSnapshot()
  })

  it('passes additional attributes to the element', () => {
    const tabs = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(tabs).toHaveProp('id', 'the-id')
    expect(tabs).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const tabs = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(tabs).not.toHaveProp('className', 'my-custom-class')
    expect(tabs).not.toHaveProp('style')
  })
})
