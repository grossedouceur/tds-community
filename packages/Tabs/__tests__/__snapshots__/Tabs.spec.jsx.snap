// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders 1`] = `
<TabsWrapper
  dataLayerEventName={true}
  tabIndex={0}
  tabsEntries={Array []}
/>
`;
